### PR TITLE
synthesize mirrors for small generic tuples

### LINKED ISSUE #7049: can't summon mirror for `MirroredElemTypes` as it is generic tuple
```scala
import scala.deriving._

case class Foo(x: Int, y: String)

def toTuple[T <: Product](x: T)(using m: Mirror.ProductOf[T], mt: Mirror.ProductOf[m.MirroredElemTypes]) =
  mt.fromProduct(x)

def weird = {
  val m = summon[Mirror.ProductOf[Foo]]
  val mt1 = summon[Mirror.ProductOf[(Int, String)]]
  type R = (Int, String)
  val mt2 = summon[Mirror.ProductOf[R]]
  val mt3 = summon[Mirror.ProductOf[m.MirroredElemTypes]]

  val f = Foo(1, "foo")
  val g: (Int, String) = toTuple(f)// (using m, mt1)
}
```

Says:

```scala
-- Error: ../pg/Main.scala:20:54 -----------------------------------------------
20 |  val mt3 = the[Mirror.ProductOf[m.MirroredElemTypes]]
   |                                                      ^
   |no implicit argument of type deriving.Mirror.ProductOf[m.MirroredElemTypes] was found for parameter x of method the in object DottyPredef
-- Error: ../pg/Main.scala:23:35 -----------------------------------------------
23 |  val g: (Int, String) = toTuple(f)// given (m, mt1)
   |                                   ^
   |no implicit argument of type deriving.Mirror.ProductOf[m.MirroredElemTypes] was found for parameter mt of method toTuple
two errors found
```

That is, we can resolve the mirror of the tuple, but not if it is obtained from another mirror. To make it work:

```scala
def weird = {
  val m = the[Mirror.ProductOf[Foo]]
  val mt1 = the[Mirror.ProductOf[(Int, String)]]
  type R = (Int, String)
  val mt2 = the[Mirror.ProductOf[R]]
  // val mt3 = the[Mirror.ProductOf[m.MirroredElemTypes]]

  val f = Foo(1, "foo")
  val g: (Int, String) = toTuple(f) given (m, mt1)
}
```

### LINKED ISSUE #14127: Tuples created through *: do not have a Mirror.Of[T]
## Compiler version
3.1.1-RC1
3.1.0

## Minimized code

```Scala
summon[Mirror.Of[Int *: String *: Boolean *: EmptyTuple]]
```

## Output

```scala
no implicit argument of type deriving.Mirror.Of[Int *: String *: Boolean *: EmptyTuple] was found for parameter x of method summon in object Predef
```

## Expectation
Output should be the same as:

```Scala
summon[Mirror.Of[(Int, String, Boolean)]]
```

since

```Scala
summon[(Int, String, Boolean) =:= (Int *: String *: Boolean *: EmptyTuple)]
```

The cause seems related to the fact that `*:` is an abstract class and those are excluded from being generic products. Relevant links:
https://github.com/lampepfl/dotty/blob/7a473349b3fb64e7c6c9dfcfb02758f4a2e3756c/library/src/scala/Tuple.scala#L318

https://github.com/lampepfl/dotty/blob/7a473349b3fb64e7c6c9dfcfb02758f4a2e3756c/compiler/src/dotty/tools/dotc/typer/Synthesizer.scala#L233

https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/transform/SymUtils.scala#L77-L88

### PR DESCRIPTION
**Pull request by bishabosha**
finally, `summon[Mirror.Of[Int *: String *: EmptyTuple]]`,

has the nice benefit of also being possible to summon a mirror that is derived from the `MirroredElemTypes` of another mirror

fixes #14127 
fixes #7049

### CODE REVIEW

**Comment by bishabosha** [comment]
need to exclude the `scala.Tuple` and `scala.*:` companion objects from themselves becoming mirrors

**Comment by bishabosha** [comment]
I think it'd probably be cleaner to make a variant of `classSymbol` that reduces to some union of classsymbol/pseudoGenericTupleClass that is ran first, and then in `acceptable` check that each part reduces to the same pseudoclass

**Comment by bishabosha** [comment]
I have simplified the way to extract the tuple proxy, with a trade off that it traverses the type twice if necessary

**Comment by bishabosha** [comment]
so removing them would require a minor release :O

**Comment by pweisenburger** [comment]
> Edit: I don't know why the erasure of Tuple.apply would change due to this PR?
> Perhaps it depends on either tupleArity or tupleElementTypes which I modified

It probably does. In PR #14586, I found the following:

```scala
def test0: Int *: EmptyTuple.type = ???  // erases to `Tuple1`
def test1: Int *: EmptyTuple = ???       // erases to `Product`
```

It doesn't seem like these types should be erased differently but they are. That PR also fixes treating `EmptyTuple` and `EmptyTuple.type` differently in general (except for erasure for binary compatibility).

**Comment by bishabosha** [comment]
> It doesn't seem like these types should be erased differently but they are. That PR also fixes treating `EmptyTuple` and `EmptyTuple.type` differently in general (except for erasure for binary compatibility).

Thanks for noticing these cases - I traced it earlier to the implementation of `tupleArity` which does not treat the widened `EmptyTuple$` ad `EmptyTuple.type`, which this PR addresses to make them be treated the same

**Comment by bishabosha** [comment]
Last changes pushed were to treat *only* `*:` as a generic tuple, before I captured arguments of `TupleN` types which had some problems with type parameters.

**Review by dwijnand — APPROVED** [review]

File: compiler/src/dotty/tools/dotc/transform/SyntheticMembers.scala [hunk 1]
```diff
@@ -26,6 +26,9 @@ object SyntheticMembers {
 
   /** Attachment recording that an anonymous class should extend Mirror.Sum */
   val ExtendsSumMirror: Property.StickyKey[Unit] = new Property.StickyKey
+
+  /** Attachment recording that an anonymous class should extend Mirror.Sum */
```
Comment (dwijnand) [inline compiler/src/dotty/tools/dotc/transform/SyntheticMembers.scala:30]: Needs fixing

Comment (bishabosha) [reply compiler/src/dotty/tools/dotc/transform/SyntheticMembers.scala]: done

**Pull request merged**